### PR TITLE
Adds autopeering

### PR DESCRIPTION
**Pull request by luca-moser**
* Adds back in the autopeering module from hive.go into Hornet.
* Autopeered peers are excluded from connection trimming and a gossip protocol is initiated with them.
* Improved the peering docs and moved them from `post_installation.md` to `peering.md`.
* Unlike legacy Hornet, entry nodes are also defined via multiaddr where a `/autopeering` portion holds the base58 encoded ed25519 public key of the entry node. Example `/ip4/127.0.0.1/udp/14626/autopeering/HmKTkSd9F6nnERBvVbr55FvL1hM5WfcLvsc9bc3hWxWc`
* Since the connection manager has no notion of a whitelist, incoming autopeering selections also initiate a connection (this seems to not cause any problems); might be changed in the future.
* The `p2pidentity` and `p2pidentityextract` tool now also output the base58 encoded public key.

### CODE REVIEW

File: pkg/shutdown/shutdown.go [hunk 1]
```diff
@@ -20,6 +20,7 @@ const (
 	PriorityKademliaDHT
 	PriorityPeerDiscovery
```
Comment (muXxer) [inline pkg/shutdown/shutdown.go]: `PriorityKademliaDHT` and `PriorityPeerDiscovery` can be removed.

File: plugins/autopeering/autopeering.go [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+
+	ipAddresses, err := iputils.GetIPAddressesFromHost(host)
+	if err != nil {
+		return nil, fmt.Errorf("unable to look up IP addresses for %s: %w", host, err)
+	}
+
+	services := service.New()
+	services.Update(service.PeeringKey, "udp", port)
+
+	ip := ipAddresses.GetPreferredAddress(false)
```
Comment (muXxer) [inline plugins/autopeering/autopeering.go]: Do we need that `CfgNetPreferIPv6` config option again maybe?

File: plugins/autopeering/local.go [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+	}
+
+	_, peeringPortStr, err := net.SplitHostPort(deps.NodeConfig.String(CfgNetAutopeeringBindAddr))
+	if err != nil {
+		log.Fatalf("autopeering bind address is invalid: %s", err)
+	}
+
+	peeringPort, err := strconv.Atoi(peeringPortStr)
+	if err != nil {
+		log.Fatalf("Invalid autopeering port number: %s, Error: %s", peeringPortStr, err)
```
Comment (muXxer) [inline plugins/autopeering/local.go]:
```suggestion
		log.Fatalf("invalid autopeering port number: %s, Error: %s", peeringPortStr, err)
```

File: plugins/autopeering/local.go [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+	}
+
+	// announce the autopeering service
+	ownServices := service.New()
+	ownServices.Update(service.PeeringKey, "udp", peeringPort)
+
+	fmt.Println("###C")
```
Comment (muXxer) [inline plugins/autopeering/local.go]:
```suggestion
```

File: plugins/autopeering/local.go [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+		ownServices.Update(p2pServiceKey(), "tcp", libp2pBindPort)
+	}
+
+	boltDb, err := bolt.CreateDB(deps.NodeConfig.String(CfgNetAutopeeringDatabaseDirPath), "autopeering.db")
+	if err != nil {
+		log.Fatalf("unable to create autopeering database: %s", err)
+	}
+
+	// realm doesn't matter
+	peerDB, err := peer.NewDB(bolt.New(boltDb).WithRealm([]byte{0}))
```
Comment (muXxer) [inline plugins/autopeering/local.go]: Can't you create the DB without `WithRealm` then?

File: config.json [hunk 6]
```diff
@@ -152,7 +152,13 @@
     "peerStore": {
       "path": "./p2pstore"
     },
-    "reconnectInterval": "30s"
+    "reconnectInterval": "30s",
```
Comment (muXxer) [inline config.json:155]: Don't forget to add these parameters to all other available config files.
Maybe we also can remove `config_as.json` ?

File: docs/post_installation/peering.md [hunk 7]
```diff
@@ -1,19 +1,115 @@
...
+More information about `multiaddr` is available at the [libp2p docs page](https://docs.libp2p.io/concepts/addressing/).
+
+### Adding node peers
+
+Once you know your node's own `multiaddr`, it can be exchanged with other node owners to establish a mutual peer
+connection. A recommended number of peer neighbors is 4-6 to get some degree of redundancy.
+
+#### Finding neighbors
+
+Join the official IOTA Discord server and join the `#fullnodes` channel and describe your node location (Europe /
```
Comment (muXxer) [inline docs/post_installation/peering.md]:
```suggestion
Join the official IOTA Discord server and join the `#nodesharing` channel and describe your node location (Europe /
```

File: plugins/autopeering/params.go [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+
+	flag "github.com/spf13/pflag"
+
+	"github.com/gohornet/hornet/pkg/node"
+)
+
+const (
+	// CfgNetAutopeeringEntryNodes list of autopeering entry nodes to use.
+	CfgNetAutopeeringEntryNodes = "p2p.autopeering.entryNodes"
+	// CfgNetAutopeeringBindAddr bind address for global services such as autopeering and gossip.
```
Comment (muXxer) [inline plugins/autopeering/params.go]:
```suggestion
	// CfgNetAutopeeringBindAddr is the bind address for autopeering.
```

File: integration-tests/tester/framework/config.go [hunk 9]
```diff
@@ -231,6 +239,50 @@ func DefaultNetworkConfig() NetworkConfig {
 	}
 }
 
+// AutopeeringConfig defines the autopeering specific configuration.
+type AutopeeringConfig struct {
+	// The ist of autopeering entry nodes to use.
+	EntryNodes []string
+	// BindAddr bind address for global services such as autopeering and gossip.
```
Comment (muXxer) [inline integration-tests/tester/framework/config.go]:
```suggestion
	// BindAddr is the bind address for autopeering.
```

File: main.go [hunk 10]
```diff
@@ -12,12 +12,12 @@ import (
 	"github.com/gohornet/hornet/core/snapshot"
 	"github.com/gohornet/hornet/core/tangle"
 	"github.com/gohornet/hornet/pkg/node"
+	"github.com/gohornet/hornet/plugins/autopeering"
 	"github.com/gohornet/hornet/plugins/coordinator"
 	"github.com/gohornet/hornet/plugins/dashboard"
 	"github.com/gohornet/hornet/plugins/debug"
 	"github.com/gohornet/hornet/plugins/migrator"
 	"github.com/gohornet/hornet/plugins/mqtt"
-	"github.com/gohornet/hornet/plugins/p2pdisc"
```
Comment (muXxer) [inline main.go:20]: You can also remove the files of this plugin and package.

File: pkg/p2p/autopeering/autopeering.go [hunk 11]
```diff
@@ -0,0 +1,226 @@
...
+	"github.com/iotaledger/hive.go/crypto/ed25519"
+	"github.com/iotaledger/hive.go/identity"
+
+	"github.com/iotaledger/hive.go/autopeering/peer"
+	"github.com/iotaledger/hive.go/autopeering/peer/service"
+)
+
+const (
+	// ProtocolCode is the protocol code for autopeering within a multi address.
+	ProtocolCode = 1337
```
Comment (muXxer) [inline pkg/p2p/autopeering/autopeering.go:24]: :carlos:

Comment (muXxer) [reply pkg/p2p/autopeering/autopeering.go:24]: Wen use `1074`

File: pkg/p2p/autopeering/autopeering.go [hunk 12]
```diff
@@ -0,0 +1,226 @@
...
+	// ErrMultiAddrNoHost gets returned if a multi address does not contain any host, meaning it neither has a /ip4, /ip6 or /dns portion.
+	ErrMultiAddrNoHost = errors.New("multi address contains no host")
+)
+
+// RegisterAutopeeringProtocolInMultiAddresses registers the autopeering protocol for multi addresses.
+// The 'autopeering' protocol value is the base58 encoded ed25519 public key and must be always 44 in length.
+func RegisterAutopeeringProtocolInMultiAddresses() error {
+	autopeeringProto := multiaddr.Protocol{
+		Name:       "autopeering",
+		Code:       1337,
```
Comment (muXxer) [inline pkg/p2p/autopeering/autopeering.go]:
```suggestion
		Code:       ProtocolCode,
```

File: pkg/p2p/discovery.go [hunk 13]
```diff
@@ -180,7 +180,7 @@ func (ds *DiscoveryService) Start(shutdownSignal <-chan struct{}) {
 // in the libp2p repository and they refer to https://github.com/libp2p/specs/pull/56.
 // however, it seems they just merged the spec without any actual implementation yet.
 func (ds *DiscoveryService) discover() {
-	discoveredPeersCount := ds.mng.ConnectedCount(PeerRelationDiscovered)
+	discoveredPeersCount := ds.mng.ConnectedCount(PeerRelationAutopeered)
```
Comment (muXxer) [inline pkg/p2p/discovery.go]: This whole file `discovery.go` can be removed.

File: plugins/autopeering/local.go [hunk 14]
```diff
@@ -0,0 +1,100 @@
...
+	multiAddrBindAddr, err := multiaddr.NewMultiaddr(bindAddr)
+	if err != nil {
+		log.Fatalf("unable to parse bind multi address %s", err)
+		return nil
+	}
+	for _, proto := range multiAddrBindAddr.Protocols() {
+		switch proto.Code {
+		case multiaddr.P_IP4:
+			peeringIP = net.IPv4zero
+		case multiaddr.P_IP6:
```
Comment (muXxer) [inline plugins/autopeering/local.go:41]: Do you maybe also need the case for `multiaddr.P_DNS` ?

**Review by muXxer — CHANGES_REQUESTED** [review]

File: config.json (hunk 6 above)
Comment (luca-moser) [reply config.json:155]: Can be done in another commit

File: plugins/autopeering/local.go (hunk 14 above)
Comment (luca-moser) [reply plugins/autopeering/local.go:41]: Not needed.

**Review by muXxer — APPROVED** [review]

**Pull request merged**